phoenix-core-server/src/main/java/org/apache/phoenix/coprocessor/PhoenixMasterObserver.java

```diff
@@ -0,0 +1,187 @@
+                String parentPartitionID = getParentPartitionId(conn, tableName, streamName, regionInfoA, regionInfoB);
+                upsertDaughterPartition(conn, tableName, streamName, parentPartitionID, regionInfoA);
+                upsertDaughterPartition(conn, tableName, streamName, parentPartitionID, regionInfoB);
+                updateParentPartitionEndTime(conn, tableName, streamName, parentPartitionID, regionInfoA.getRegionId());
+            }
+        } catch (SQLException e) {
+            LOGGER.error("Unable to update CDC Stream Partition metadata: " + e);
+        }
+    }
+
+    private PTable getPhoenixTable(Connection conn, TableName tableName) throws SQLException {
+        PTable pTable;
+        try {
+            pTable = PhoenixRuntime.getTable(conn, tableName.toString());
+        } catch (TableNotFoundException e) {
+            return null;
+        }
+        return pTable;
+    }
+
+    /**
+     * Lookup parent's partition id (region's encoded name) in SYSTEM.CDC_STREAM.
+     * We infer start and end key for the parent from the two daughters' start/end keys.
+     */
+    private String getParentPartitionId(Connection conn, String tableName, String streamName,
+                                        RegionInfo regionInfoA, RegionInfo regionInfoB)
+            throws SQLException {
+        byte[] inferredParentStartKey =
+                Bytes.compareTo(regionInfoA.getStartKey(), regionInfoB.getStartKey()) < 0 ?
+                regionInfoA.getStartKey() : regionInfoB.getStartKey();
+
+        byte[] inferredParentEndKey = (regionInfoB.getEndKey().length == 0 || regionInfoA.getEndKey().length == 0)
+                ? new byte[0]
+                : Bytes.compareTo(regionInfoA.getEndKey(), regionInfoB.getEndKey()) < 0
+                    ? regionInfoB.getEndKey()
+                    : regionInfoA.getEndKey();
+
+        StringBuilder qb = new StringBuilder(PARENT_PARTITION_QUERY);
+        if (inferredParentStartKey.length == 0) {
+            qb.append(" AND PARTITION_START_KEY IS NULL ");
+        } else {
+            qb.append(" AND PARTITION_START_KEY = ? ");
+        }
+        if (inferredParentEndKey.length == 0) {
+            qb.append(" AND PARTITION_END_KEY IS NULL ");
+        } else {
+            qb.append(" AND PARTITION_END_KEY = ? ");
+        }
+
+        PreparedStatement pstmt = conn.prepareStatement(qb.toString());
+        int index = 1;
+        pstmt.setString(index++, tableName);
+        pstmt.setString(index++, streamName);
+        if (inferredParentStartKey.length > 0) pstmt.setBytes(index++, inferredParentStartKey);
+        if (inferredParentEndKey.length > 0) pstmt.setBytes(index++, inferredParentEndKey);
+        LOGGER.info(pstmt.toString());
+        ResultSet rs = pstmt.executeQuery();
+        if (rs.next()) {
+            return rs.getString(1);
+        } else {
+            throw new SQLException("Could not find parent of the provided daughters.");
```

Comment:
Add regionInfoA and regionInfoB start/end keys in the Exception message?

Reply:
done